docs/design/auditing-containment.md

```diff
@@ -0,0 +1,205 @@
+  piece_id bytea NOT NULL,
+  piece_num bigint NOT NULL,
+  stripe_index bigint NOT NULL,
+  share_size bigint NOT NULL,
+  expected_share_hash bytea NOT NULL,
+  reverify_count integer NOT NULL,
+  PRIMARY KEY ( node_id, segment_path )
+)
+```
+
+## Rationale
+
+Originally I had the idea to have a separate service iterate through and check the ContainmentDB.
+However, after talking with JT it seemed unnecessary to have another service for this functionality that could be added to the existing verifier code.
+I think we also don't want to pound the contained nodes with reverification audits.
+They should just be reverified whenever they pass through the audit service normally.
+
+## Implementation
+
+#### 1. Uncooperative nodes are documented in the ContainmentDB:
+
+By "uncooperative," I mean that the verifier will likely be able to successfully DialNode (within pkg/audit/verifier.go), but errors may occur at
+```go
+downloader, err := ps.Download(timedCtx, limit.GetLimit(), offset, int64(shareSize))
+```
+or
+```go
+_, err = io.ReadFull(downloader, buf)
+```
+Once these errors have occurred and we know that a `pendingAudit` entry should be made in the ContainmentDB, we should use infectious’s `f.Decode` to generate the original stripe.
+Then we should use infectious’s `f.EncodeSingle` where we input the stripe and output the missing share. We then make a hash of that missing share to be saved to the ContainmentDB as `ExpectedShareHash`.
+
+`containment.IncrementPending` should be called when a node is found to be “uncooperative."
+
+```go pkg/audit/verifier.go
+err := verifier.containment.IncrementPending(ctx, &pendingAudit{
+    nodeID:            nodeID,
+    pieceID:           pieceID,
+    pieceNum:          pieceNum,
+    stripeIndex:       stripeIndex,
+    shareSize:         shareSize,
+    expectedShareHash: shareHash,
+}
+```
+
+#### 2. Audits continue normally. But now the nodes listed in a pointer are checked for `contained` status.
+```go pkg/audit/containment/checker.go
+func (verifier *Verifier) Verify(ctx context.Context, stripe *Stripe) (verifiedNodes *RecordAuditsInfo, err error) {
+  defer mon.Task()(&ctx)(&err)
+
+  pointer := stripe.Segment
+
+  pieces := pointer.GetRemote().GetRemotePieces()
+
+  // Get the NodeDossiers from the overlay using the pieces’ nodeIDs.
+  // Determine if they have the contained flag set to true.
+  // If so, call Reverify on them.
+  // Either way, then continue this audit normally with the contained nodes included.
+  ...
+}
+```
+
+#### 3. Contained nodes are checked for the same data that they were originally requested to respond with:
+
+```go pkg/audit/verifier.go
+func (verifier *Verifier) Reverify(ctx context.Context, node storj.NodeID) (auditRecords *RecordAuditsInfo, err error) {
+
+  pendingAudit, err := checker.containment.Get(id)
+
+  // Dial the node, then use info from the pendingAudit to download the target share from the node.
+
+  offset := node.stripeIndex * node.shareSize
```

Comment:
this info should come from `pendingAudit`, not `node`, right?

Reply:
yeah good catch